### ISSUE TITLE
join_return_with_assignment false positive

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/linter#1457]
**Describe the issue**

join_return_with_assignment reports an error for some cases where the assigned variable is used inside of the same assignment block (generally in a lambda).

**To Reproduce**

Here is a somewhat contrived example but similar to the case I ran into:

```
Stream foo() {
  StreamController controller;
  controller = StreamController(onListen: () {
    // controller needs to be assigned or you get a null ptr exception here.
    someOtherStream.listen(controller.add);
  });
  return controller.stream;
}
```

**Expected behavior**

The lint would not warn about this code since controller is used after assignment (although its admittedly a bit difficult to tell here).

**Comment by pq** [comment]
Thanks for this repro @jakemac53 !

**Comment by srawlins** [comment]
I think this might be very tricky to implement. For example assigning to a field requires care:

```dart
class C {
  late StreamController controller;
Stream foo() {
  controller = StreamController(onListen: () {
    // controller needs to be assigned or you get a null ptr exception here.
    someOtherStream.listen(controller.add); // _this_ controller.
    someOtherStream.listen(otherC.controller.add); // But _not this_ controller.
  });
  return controller.stream;
}
```
